### PR TITLE
OTEL: Dynamic service name for tracing in K8s environment

### LINKED ISSUE #32095: OTEL: Dynamic service name for tracing in K8s environment
### Description

Keycloak tracing capabilities now support changing the service name of the exported traces. For K8s environment, it'd be great to have the service name dynamic in order to differentiate the traces between different deployments (as stated by @ahus1 in https://github.com/keycloak/keycloak/issues/32056). At the moment, the service name is fixed to be "keycloak", and the host name in a Kubernetes setup would always be "keycloak-0" / "keycloak-1" / ...

The most suitable approach for it would be leveraging [Downward API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/)

### PR DESCRIPTION
**Pull request by mabartos**
Closes #32095 

k9s: env vars
image

Pod log:
```
resource=Resource{schemaUrl=null, attributes={host.name="example-kc-0", k8s.namespace.name="keycloak-test-14b6b809-0814-4255-a910-f34b1684e965", service.name="example-kc", ...
```

It could be good to provide some test cases by analyzing the pod log and the used service name - but not sure how. @vmuzikar @shawkins Is it possible to achieve it somehow?

### CODE REVIEW

**Comment by mabartos** [comment]
Ooops, the added test failed.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank for this first use of the downward API. I see that the service name now matches the pod name, still I think this is not what users will have in mind: In the example you provided, the service name equals the pod name, which is usually the host name in a Kubernetes setup. 

My original comment mentioned that people would want to differentiate deployments between namespaces. What I forgot to mention was that there can be multiple deployment in a single namespace. 

To differentiate those, users would need on each trace the information about namespace (not included here), service (in your case "example-kc", and it is shared by all pods of a deployment), and pod (can be the host name as before, no extra attribute needed IMHO). AFAIK the service name is available via the label `app.kubernetes.io/instance` in the downward API, still I might be wrong here and you might choose to pass it directly via the Operator.

**Comment by mabartos** [comment]
@ahus1 You're right; I got a little bit mixed up with different deployments terms. I'll comprehend the capabilities of the tracing in k8s env more for better filtering of traces.

**Comment by mabartos** [comment]
@ahus1 Updated. Filtering based on pod name, deployment name, and namespace. We did not have to use the Downward API for such simple cases. 

There are different attribute names to obtain such information, but I'd say it's better to document it than duplicating these values with unified `k8s.*` attributes for OTel:
```
service.name 
k8s.namespace.name
host.name
```

WDYT?

**Review by ahus1 — DISMISSED** [review]
Thank you for this change. Looking at the docs and also the tests this looks all good to me.

**Comment by ahus1** [comment]
Not merging yet as you might be waiting for a review from the CND team as well.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java
Comment (shawkins) [inline operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java:438]: Not a problem for what's here, but just wanted to double check something. Looking at the well known key / values if you wanted to add annotations those values can contain commas, which would likely throw off the parsing. Is there escaping for commas, or would attributes like that not be supported?

**Review by shawkins — COMMENTED** [review]
This looks good, just a general question about the attributes.

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java
Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java:438]: @shawkins I'm not sure if I understand. Could you elaborate a little bit more?

Are you referring to the resource attributes specified in the env var and then parsing on the server side? The parsing of these comma-separated values should be covered in this test case: https://github.com/keycloak/keycloak/blob/main/quarkus/tests/integration/src/test/java/org/keycloak/it/cli/dist/TracingDistTest.java#L176

**Review by shawkins — DISMISSED** [review]
LGTM

File: docs/guides/server/tracing.adoc
Comment (vmuzikar) [inline docs/guides/server/tracing.adoc:125]: `additionalOptions` is a supported way to configure Keycloak
```suggestion
NOTE: There is no support for tracing configuration in {project_name} CR yet, so the `additionalOptions` can be used to the `tracing-enabled` property and other tracing options.
```

File: operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java:438]: There will be currently always only a single entry `k8s.namespace.name={namespace}` if I understand correctly. So no need to handle escaping the commas?

Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java:438]: @vmuzikar correct, the namespace value will not contain , nor \

Comment (mabartos) [reply operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java:438]: @shawkins Good point. TBH, I haven't further investigated such a "corner" case and rather have been relying on the Quarkus handling as these attributes are a list of items in the format `key1=val1,key2=val2,key3=val3`[1]. 

I think it's sufficient ATM.

> There will be currently always only a single entry k8s.namespace.name={namespace} if I understand correctly. So no need to handle escaping the commas?

For the future, we might consider adding other attributes besides the namespace. 

[1] https://quarkus.io/guides/opentelemetry#quarkus-opentelemetry_quarkus-otel-resource-attributes

**Review by vmuzikar — APPROVED** [review]
@mabartos LGTM, thanks for these changes.

**Pull request merged**
